Update `ArrayBuffer` types to support TypeScript 5.9

jamesbvaughan opened the pull request:
When trying to upgrade a project from TypeScript v5.8 to v5.9, I hit the issue described here: https://devblogs.microsoft.com/typescript/announcing-typescript-5-9/#lib.d.ts-changes

This implements the suggested fix described there in order to make the library compatible with TypeScript 5.9.

donmccurdy commented:
Thanks @jamesbvaughan! I'm wondering what to do with other Uint8Array references moving forward. Assuming that every Uint8Array used in the library is already a `Uint8Array<ArrayBuffer>`, does this sound like a reasonable plan for the TypeScript update?

- Functions returning a Uint8Array should change their return type to `Uint8Array<ArrayBuffer>`
- Functions accepting a Uint8Array as a parameter should continue to accept `Uint8Array` (no change) for backward-compatibility

jamesbvaughan commented:
I had targeted this specific function because it's the one that was blocking our upgrade, but I'll take some time today to make this more complete.

jamesbvaughan commented on packages/functions/src/texture-compress.ts:
```diff
@@ -338,7 +338,7 @@ async function _encodeWithSharp(
 	}
 
-	return BufferUtils.toView(await instance.toBuffer());
+	return BufferUtils.toView((await instance.toBuffer()) as Buffer<ArrayBuffer>);
```
This either requires a type assertion or an update to `sharp`'s types.

jamesbvaughan commented:
Alright I've updated the types more comprehensively now.

My process was to update TypeScript in this repo to 5.9, then to repeatedly run the build and fix errors until I addressed every type error that was breaking the build.

I didn't include the TypeScript upgrade in this branch but I can if you'd like me to.

A few notes:
- I only made as many changes as were necessary to get the whole repo to build (with `yarn build`) without type errors under TypeScript 5.9.
- In some situations where 3rd party libraries didn't have the required type updates, I opted to us `as FooArray<ArrayBuffer>`.

I'd understand if you would prefer not to introduce those type assertions and want to hold off on these changes until all of the relevant dependencies have updated.

donmccurdy commented on packages/view/src/ImageProvider.ts:
```diff
@@ -66,13 +66,13 @@ export class DefaultImageProvider implements ImageProvider {
 		const mimeType = textureDef.getMimeType();
 
-		let texture = this._cache.get(image);
+		let texture = this._cache.get(image.buffer);
```
Hm, I should be using Uint8Array here, but that's probably just something I missed in refactoring a bit ago. Will clean this one up in a separate PR.

donmccurdy commented:
Thanks @jamesbvaughan! The changes look good to me, I'll merge shortly. In general I think I'd like for the type definitions to be (in this case) loose in what they accept (e.g. `Uint8Array`) but strict in what they return (e.g. `Uint8Array<ArrayBuffer>`), which is consistent with your changes, it looks like.

Updates to upstream packages may be less trivial, for example mikktspace-wasm's types are built by wasm-bindgen I believe.
